server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java

```diff
@@ -68,147 +66,171 @@
+        } else {
+            return Either.left(
+                    new ErrorWithId(
+                            entry.getCreationId(), 
+                            SetError.builder().type("error").description("Not yet implemented").build()));
+        }
+    }
+    
+    private Either<ErrorWithId, CreationMessageEntry> getInvalidArgumentsErrors(CreationMessageEntry entry) {
+        CreationMessage message = entry.getValue();
+        if (!message.isValid()) {
+            return Either.left(new ErrorWithId(entry.getCreationId(), buildSetErrorFromValidationResult(message.validate())));
+        }
+        return Either.right(entry);
     }
+    
+    private Either<ErrorWithId, MessageWithId> handleOutboxMessages(CreationMessageEntry entry, MailboxSession session) {
+        CreationMessageId creationId = entry.getCreationId();
 
+        return getMailboxWithRole(session, Role.OUTBOX)
+            .map(outbox -> {
+                if (isRequestForSending(entry.getValue(), session)) {
+                    Function<Long, MessageId> idGenerator = uid -> generateMessageId(session, outbox, uid);
+                    return createMessageInOutboxAndSend(entry, session, outbox, idGenerator);
+                }
+                return Either.<ErrorWithId, MessageWithId>left(
+                        new ErrorWithId(creationId, 
+                                SetError.builder().description("unhandled request").type("unhandled").build()));
+                
+            }).orElse(Either.left(new ErrorWithId(creationId, UNEXPECTED_ERROR)));
+    }
+    
     @VisibleForTesting
-    protected MessageWithId<Message> createMessageInOutboxAndSend(MessageWithId.CreationMessageEntry createdEntry,
+    protected Either<ErrorWithId, MessageWithId> createMessageInOutboxAndSend(CreationMessageEntry createdEntry,
                                                            MailboxSession session,
                                                            Mailbox<Id> outbox, Function<Long, MessageId> buildMessageIdFromUid) {
+        CreationMessageId creationId = createdEntry.getCreationId();
         try {
             MessageMapper<Id> messageMapper = mailboxSessionMapperFactory.createMessageMapper(session);
             MailboxMessage<Id> newMailboxMessage = buildMailboxMessage(createdEntry, outbox);
             messageMapper.add(outbox, newMailboxMessage);
             Message jmapMessage = Message.fromMailboxMessage(newMailboxMessage, buildMessageIdFromUid);
             sendMessage(newMailboxMessage, jmapMessage, session);
-            return new MessageWithId<>(createdEntry.getCreationId(), jmapMessage);
-        } catch (MailboxException | MessagingException | IOException e) {
-            throw Throwables.propagate(e);
+            return Either.right(new MessageWithId(creationId, jmapMessage));
+            
+        } catch (MessagingException | MailboxException e) {
+            return Either.left(new ErrorWithId(creationId, UNEXPECTED_ERROR));
```

Comment:
I find Either being really poor here, as we are losing entirely the content of the exception.